1799: Flatten data fetching

kael89 opened the pull request:
### Issue #:
Actual implementation of https://github.com/beyondessential/tupaia-backlog/issues/1799 (a couple of previous PRs were just moving things around to make the change cleaner)

**Note:** tests here will fail as I haven't updated them yet. I thought it's better to get a review on the code and most importantly the architecture in parallel to writing the tests, in case major changes are requested.

### Architecture

#### Algorithm
1. `IndicatorApi` asks each indicator which data elements they use
2. If a data element is of `indicator` type, we pull it from the `indicators` table and ask it for its data elements. If not, we add it in a list of codes to be fetched
3. This is done one nesting level at a time. We define a max nesting depth, since indicators may (erroneously) be self-referencing or cyclical (A => B => A)
4. After we retrieve all required information, we fetch all data elements at once. The results are passed in all indicators to do with them as they please. No aggregation is applied.
5. Indicators aggregate the fetched data post-fetch.

#### Data fetching and aggregation

In some cases, data fetching and subsequent aggregation are coupled. This happens when the aggregation needs data outside the originally requested date range (or entities, in the future). For example, if we want to aggregate the results using `OFFSET_PERIOD` for -1 weeks, we need to:
* Add -1 week when querying the DB (fetch step)
* Add +1 week when returning the results (aggregation step)

Fetch and aggregation are now decoupled, which makes the correct and elegant coordination of the above trickier. We need to share some state between the fetch and aggregation steps, so that the first can be done based on the needs of the second. I tried to encapsulate the "expanding" and "shrinking" of date ranges in the `AnalyticsRepository` class. I tried to hide as many details as possible, but there is still a bit of abstraction leaking (compare `aggregateRootAnalytics` vs `aggregateNestedAnalytics`).

I believe that this is a result of the underlying separation of concerns here, but let me know if you think of a better pattern.

kael89 commented on packages/indicators/src/IndicatorApi.ts:
```diff
@@ -26,39 +29,91 @@ export class IndicatorApi implements IndicatorApiInterface {
+    const indicatorCodes: string[] = [];
+    const nonIndicatorCodes: string[] = [];
+
+    const dataElements = await this.models.dataSource.findOrDefault({
```
This is a bit dodgy here: we depend on `DataSource.findOrDefault`, which is specifically defined in the `DataSource` model in `@tupaia/data-broker`, and not in the base model in `@tupaia/database`.

It works in practice since `IndicatorApi` is currently instantiated in `DataBroker`, which passes in its own models. https://github.com/beyondessential/tupaia-backlog/issues/663 should solve this issue. Until then, we could approach this in various ways but I thought that leaving as is will be most probably fine in practice.

edmofro commented on packages/indicators/src/AnalyticsRepository.ts:
```diff
@@ -0,0 +1,128 @@
+      expandedFetchOptions,
+    );
+
+    this.data = {
+      analyticsByElement: groupBy(analytics, 'dataElement'),
+      fetchOptions,
+      expandedFetchOptions,
+    };
```
I would tend to store each of these three as individual fields - later it confuses me a bit to see `this.data.fetchOptions` etc.

kael89 replied:
Of course you are right! Naming a variable `data` is definitely a smell. Will change

kael89 replied:
I realise now why I stored all of them under a `data` field: because we need to check that the repository is populated before attempt to access any of those fields.

I could still add them as separate fields with a private getter/setter for each, but it's more verbose and it makes the distinction between populated/non-populated fields less clear. 

Based on the above, I'd prefer this to stay as is - I don't know if it's defined as such but in my head I'd call it the "Repository" pattern.

edmofro replied:
Right....what about renaming the field to something like `populatedFields`?

kael89 replied:
I guess fields is more specific than  "data"

edmofro replied:
And less easily confused with the actual analytic data (which is the mistake I made when reading through the first time).

Thanks!

kael89 replied:
I did the rename in [da4a989](https://github.com/beyondessential/tupaia/pull/1873/commits/da4a98914c7f424d8879ebb5d5e225ad4fd0132d): essentially I renamed `data` to `fields`.

Field access is done like this:
```ts
this.fields.fetchOptions
```

You may had the following in mind:
```ts
this.populatedFields.fetchOptions
```

But then I wouldn't know how to name the actual field in the class. Currently it is
```ts
private populatedFields: Fields | null = null;
```

and accessed via a getter/setter.

Happy to further rename it if you have another idea!

edmofro replied:
No all good! You're right about my expectation but I saw that and realised why and am very happy with `fields` :-)

edmofro commented on packages/indicators/src/AnalyticsRepository.ts:
```diff
@@ -0,0 +1,128 @@
+   * This method should be used for analytics in nested indicators.  Expanded fetch options are used,
+   * to allow calculations that may require data outside the originally requested dimensions
+   */
+  aggregateNestedAnalytics = (analytics: Analytic[], aggregations: Aggregation[]) =>
```
It really took a while for me to get my head around what aggregateRootAnalytics vs. aggregateNestedAnalytics meant. I don't really think there are better names, but if you have any thoughts in improving dev understanding here, let me know

kael89 replied:
Yes I believe that's the weakest point of my implementation... But it's the best version I ended up with.

So the comments above them didn't help? Should we just accept that the shrink/expand of fetchOptions (and resulting need for a different nested/root logic) is just complicated business, or should we brainstorm ways to make this even simpler?

edmofro replied:
The comments helped, but didn't get me all the way there. I think it might be nice to illustrate the flow somehow, because I struggled to understand the full flow. Tbh I still do - here's my current understanding:
- nested analytics will already have been aggregated further down the chain by their own indicator builder
- that indicator builder will have aggregated them using the dates, as when it aggregates its own analytics it will do so using the root aggregation process
- at this point the root indicator builder is applying another layer of aggregation on top of the nested analytics that have been passed up, and shouldn't apply the tight date constraints as the dates will have been sorted out by the nested builders

Even in writing that I'm not sure it's true. Question:
- Is it the case that each indicator builder will be considered the "root" at some point, or is there just one root per fetch?

kael89 replied:
There is only one root per fetch

edmofro reviewed (changes requested):
Very nice work @kael89! I've made some specific comments, but I have a couple of more general questions that would be more suitable to a call:
- ~How does the nested aggregation make sure it's got correctly constrained data~ have now understood this, I think
- The ArithmeticBuilder seems to be doing a lot of work of creating builders and getting their data elements etc that I would have expected to happen outside of that class...but perhaps there are strong ties to the structure of the arithmetic builder's config

kael89 commented on packages/indicators/src/types.ts:
```diff
@@ -42,31 +41,30 @@ export type Indicator = {
 export type IndicatorType = DatabaseType<IndicatorFields>;
 
+type DataSourceModel = DatabaseModel<DataSourceFields> & {
+  findOrDefault: DatabaseModel<DataSourceFields>['find'];
```
Exactly! It would be the same as `DatabaseModel.find` in JavaScript. It's a way to reference properties of already defined types/interfaces

kael89 commented on packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts:
```diff
@@ -0,0 +1,184 @@
+  getAggregations = (): Aggregation[] => {
+    const formulaAggregations = Object.values(this.config.aggregation).flat();
+    const parameterAggregations = this.config.parameters
+      .map(p => createBuilder(p).getAggregations())
```
Great point @edmofro ! I also thought I could introduce a `parameterBuilderCache` (similar to the `configCache` in this class). But I have the impression that this operation is fairly cheap anyway.

I may attempt to do so and get an actual measurement. I suspect the performance diff will be negligible (although you never know!), but it may even make the code a bit simpler!

kael89 replied:
Eventually I did use a parameter builder cache, since it seems to simplify code!

kael89 commented on packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts:
```diff
@@ -0,0 +1,184 @@
+   * b. Nested indicators
+   * c. "Primitive" elements (eg `dhis`, `tupaia` elements)
+   */
+  private buildAggregatedAnalytics = (
```
Totally! Although it felt tempting, I decided to not to do so until the need arises

kael89 commented on packages/aggregator/src/analytics/aggregateAnalytics/adjustTemporalDimensionsToAggregation.js:
```diff
@@ -59,3 +59,27 @@ export const adjustTemporalDimensionsToAggregation = (temporalDimensions, aggreg
+ * to be adjusted, eg when we need to take into account past data
+ * Return the widest values for each adjustable fetch option
+ */
+export const expandFetchOptionsToSpanAggregations = (fetchOptions, aggregations) => {
```
Thanks, we should definitely unit test this!

kael89 commented on packages/utils/src/object.js:
```diff
@@ -192,3 +194,13 @@ export const filterValues = (object, valueFilter) =>
+
+const sortFields = object => {
+  const sortedEntries = Object.entries(object).sort(([keyA], [keyB]) => compareAsc(keyA, keyB));
+  return Object.fromEntries(sortedEntries);
```
I don't understand this but maybe you could explain it with an example? Or in a call

edmofro replied:
@mcccclean can you confirm whether I'm right here? 

Starting with two objects with the same key/value pairs (none of which are nested objects) but in undefined orders
- get the entries out of the object
- sort them by key
- construct a new object from those sorted entries
- JSON.stringify the object
If you then compare the results, are they guaranteed to match? I am thinking no, because when you construct the new object from the sorted entries, the order of the entries in the new object is not guaranteed?

kael89 replied:
@edmofro perhaps you refer to [Object.entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries)? The docs for it mention

> The order of the array returned by Object.entries() does not depend on how an object is defined

---

For `Object.fromEntries`:

According to its [specification](https://tc39.es/ecma262/#sec-object.fromentries), it uses the [AddEntriesFromIterable](https://tc39.es/ecma262/#sec-add-entries-from-iterable) algorithm to create the new object. My understanding is that this will result in properties to be inserted in the new object in the same order they are defined in the entries array.

On the other hand, `JSON.stringify` doesn't guarantee property order (see [this note](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#issue_with_plain_json.stringify_for_use_as_javascript)).  However, I would assume that for an object with the same props defined in the same order we'll always get the same string, even if the fields are re-ordered in the stringified result. (This is not relevant directly to this function, but it is to [getUniqueObjects](https://github.com/beyondessential/tupaia/blob/1779-flatten-data-fetching/packages/utils/src/object.js#L203) which uses it below)

edmofro replied:
My understanding was just that in general, objects never guarantee order of properties

edmofro replied:
However if you're using the same exact operations to construct and then stringify the object, I agree that you would expect that to be deterministic, and the result should be the same (even if the order isn't the same as the order they were sorted into)

kael89 replied:
Yes, that would work for `getUniqueObjects`. However, `sortFields` could still be a misnomer for this function, if the enumeration order does not match our intended field sort order.

I believe in this specific case this happens to be fine though. We use `compareAsc` for sorting, which will place all integers sorted by value first (1 before 10 etc), then all alphanumerics sorted in ASC order. This matches the enumeration order introduced in ES6 according to [this article](https://dev.to/eladtzemach/traversal-order-of-object-properties-in-javascript-es6-5a3) (haven't found an official doc yet!)

So while in the general case the insertion order is not preserved during enumeration, in this case our sorting logic just happens to be compatible with the prop enumeration logic, which is a happy coincidence 😆

mcccclean replied:
Weighing in - my understanding is that property order in an object [_is_ predictable](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/), but only since ES2015, and that that assumption might not survive transpilation to an earlier spec? So if this is getting transpiled to support older browsers it might break (and it would break on newer browsers too, because it'd be running the same transpiled code as an old IE or whatever).

BUT, the transpilation process is itself predictable! So, like, given that this particular code is running correctly now, I would feel pretty safe to assume that it runs correctly in general.

edmofro replied:
Cool cool, it sounds like we all agree that this is totally fine in practice :-)

kael89 replied:
That was an interesting discussion! It made me realise that we don't handle string and number comparison correctly in `compareAsc`. We used `<` and `>`, which always return `false` when a string and a number are compared, regardless of the operand order.

This is now fixed in the latest [commit](https://github.com/beyondessential/tupaia/pull/1873/commits/6148c3744ff7cac4dc832b67f27fa3d62a707bee). I also added a compatibility note in `sortFields`

kael89 commented on packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts:
```diff
@@ -0,0 +1,184 @@
+    const buildAnalyticsUsingBuilder = (builder: Builder) =>
+      builder.buildAnalytics(populatedAnalyticsRepo, buildersByIndicator, fetchOptions);
+
+    if (isParameterCode(parameters, variable)) {
+      const parameter = parameters.find(p => p.code === variable) as Indicator;
```
That's a good point and a classic dilemma when working with linear structures, I guess! I'm thinking I can keep this optimised and readable by using a `getParameter(code`) helper

kael89 commented on packages/indicators/src/types.ts:
```diff
@@ -42,31 +41,30 @@ export type Indicator = {
 }
 
-export type FetchOptions = Readonly<Record<string, unknown>>;
+export type FetchOptions = Readonly<{ startDate: string; endDate: string; period?: string }>;
```
`startDate` and `endDate` are mandatory since they are required by `@tupaia/data-api`, which doesn't support `period... `period` can also be passed in since DHIS supports it, but should probably be deprecated

kael89 commented:
@edmofro I have addressed all your comments except than the one we talked in our call. I'll try to phrase it here for the record:

The current implementation uses a naive approach which expands `fetchOptions` for all nested aggregator calculations. This won't play nicely with aggregations like `SUM`, whose range is implicitly specified via the provided `fetchOptions`. Suggested solution:
  * Calculate the most expanded `fetchOptions` to fetch all data with one request (we do so already)
  * Calculate whether `fetchOptions` (=> "aggregation options" at this stage) should be expanded or not per each aggregation. A few aggregation types (`OFFSET_PERIOD` and `SUM_UNTIL_CURRENT_DAY`) will need to expand options, but most of the current aggregation will (and should!) not

I was thinking to finish writing the tests in a new PR, then add a failing test case for the above, then make it pass. No need to approve this PR till then

kael89 commented on packages/indicators/src/Builder/ArithmeticBuilder/config/aggregation.ts:
```diff
@@ -79,12 +78,12 @@ const validateAggregationDictionary = (
   const parser = getExpressionParserInstance();
   parser.getVariables(formula).forEach(variable => {
-    if (!(variable in aggregation) && !isParameterCode(parameters, variable)) {
+    if (!(variable in aggregation)) {
```
In the past, we didn't throw an error if a parameter did not have an aggregation specified for it. This was because we had no way of aggregating parameter results. Aggregation was coupled with data fetching in Aggregator. which in turn only works with data elements (eg indicators) defined in the DB, and not for inline indicators (parameters).

Now that aggregation can be done independently from fetching, we can aggregate parameter results, which may actually prove handy.

edmofro reviewed (commented):
Nice - I really like the parameter builder cache as it pulls the knowledge of builder construction into a single place (which could potentially be moved around in future) and simplifies the logic of the code further down (as well as a minor performance benefit!)

kael89 commented:
Hey @edmofro - do you mind approving this PR, and then I can address all remaining items in one last PR against the base [1779-optimise-data-fetching-for-indicators](https://github.com/beyondessential/tupaia/tree/1779-optimise-data-fetching-for-indicators) branch?

Still TODO:
[ ] fix [aggregation logic](https://github.com/beyondessential/tupaia/pull/1873#issuecomment-761892449)
[ ] confirm that `getUniqueObjects` in ArithmeticBuilder.getElementCodes doesn't have significant performance impact
[ ] try using a `.js` file instead of `.json` for defining integration test config

edmofro reviewed (approved):
Sure :-)
